### ISSUE TITLE
Castle Editor: Component name is not updated in TreeView if "Same name" error pops up

### ISSUE THREAD

**Issue by eugeneloza** [castle-engine/castle-engine#411]
It just becomes blank (however "hint" is correct - the previous name of the component before error message)

![CGE-name-missing-after-rename-into-same-name](https://user-images.githubusercontent.com/10726162/201228984-b9b88bd7-660b-473b-b886-3b02bcd6434a.png)

**Comment by michaliskambi** [comment]
Hm, I cannot reproduce this -- 

> 2. Add new element with the name exactly the same as the root element

This step seems impossible, i.e. if you add a components, it will always have a unique name.

Note: In case you're making custom components that create their own children:

1.In general, these children should be internal, which means

    - owner of the child should be the creator (like `TMyInternalComponent.Create(Self)`)
    - call `.SetTransient` on child
    - name of child can be whatever, it will not conflict with names in the design, as the components in the design have a different owner (`DesignOwner` in editor code)
    - E.g.: see how TCastleCheckbox creates internal components, src/ui/castlecontrols_checkbox.inc , like FImageContainer, FImagePressedBackground etc.

2. If you don't want these children to be internal, then

    - owned of the child should be owner of the component (like `TMyInternalComponent.Create(Owner)`)
    - set name of the child using `InternalProposeName`, like `Light.Name := InternalProposeName(TCastlePointLight, Owner);`
    - E.g. see TCastleViewport.SetupChildren3D in src/scene/castleviewport.pas

And yes, it should be documented in https://castle-engine.io/custom_components , I'll add it there soon :)

**Comment by eugeneloza** [comment]
I can always easily reproduce this issue on Debian 10 and 11. It doesn't have anything to do with custom components - the most basic empty design in an empty project has this issue. Maybe it has something to do with XFCE and dark theme?

https://user-images.githubusercontent.com/10726162/202124620-9c7c990f-b04a-4677-a62e-f5b1eadc6650.mp4

Note that my CGE installation is a bit obsolete (by a couple of weeks most likely). I can update it to make sure it wasn't a consequence of some at first glance unrelated but already fixed issue.

**Comment by michaliskambi** [comment]
Hmm, the same sequence works OK for me. I tried to do exactly what you show, I initiated rename by F2 or by clicking, and it always goes OK. Testing on Ubuntu 22.10 now, with GNOME.

Note that name reverts to `Design1` (it actually never changed to `Group1`) already when you see the "Duplicate name" error.

Indeed please upgrade and retest. Looking at your movie, maybe it's indeed some theme-specific issue -- as it seems there is a horizontal space for `Design1` name, it's just not shown for some reason. If the issue persists after upgrade, I will of course want to fix/workaround it in the editor anyway (no matter whether it's Xfce specific problem or not).


https://user-images.githubusercontent.com/1076978/202288293-8a102723-b894-43b2-9502-e4a543707b03.mp4

**Comment by michaliskambi** [comment]
> Note: In case you're making custom components that create their own children:

@eugeneloza 

I have extended the https://castle-engine.io/custom_components page with a section "9. Adding children components" with lots of text and 2 examples how to add children to your components. I hope that this is useful :)

( I know that it is no longer related to this issue, i.e. I understand you can reproduce it without any custom components. Still, I know you're doing custom components, so I hope that new section will be useful to you and it was inspired by my first answer here. )

**Comment by eugeneloza** [comment]
Reading through:

![2022-11-17 09_35_04-Greenshot](https://user-images.githubusercontent.com/10726162/202384508-289c6c1c-ebbf-42a3-831b-302b2533c0c6.png)

Hmmm... First of all, it seems like `constructor` is not called after deserealization and it seems like an universal feature. I've had a lot of problems with that in City Stories. I may be wrong here though.

Second, indeed I create the children in `constructor` (the slider knob in the sliders asset) and it seems to work fine. I was mimicking how it's done in `TCastleScrollView`: https://github.com/castle-engine/castle-engine/blob/master/src/ui/castlecontrols_scrollview.inc#L598 - Though I already see that I didn't set `FScrollArea.SetSubComponent(true);`: looking through CGE documentation doesn't give much information:

![2022-11-17 09_41_24-Greenshot](https://user-images.githubusercontent.com/10726162/202385586-7fd2267f-476d-48d5-99a8-2344bc16214d.png)

I'll look at the code a bit later.

**Comment by eugeneloza** [comment]
![2022-11-17 09_54_42-Greenshot](https://user-images.githubusercontent.com/10726162/202388182-958b5f1e-daaf-468f-8875-ebea496bc442.png)

Also I see that ScrollView also sets a static name for `ScrollArea`: https://github.com/castle-engine/castle-engine/blob/master/src/ui/castlecontrols_scrollview.inc#L600 

I reimplemented `InternalProposeName` for my goals (if I remember correctly, because it was crashing), but then rolled the change back when I saw that having the "static" name in `ScrollArea` is ok and it seems like not causing any troubles in my usecase (but of course doesn't look clean)

**Comment by michaliskambi** [comment]
> Hmmm... First of all, it seems like constructor is not called after deserealization and it seems like an universal feature. I've had a lot of problems with that in City Stories. I may be wrong here though.

The constructor is always called (in case of CGE and Pascal in general). And constructor is called before deserializing, i.e. before the properties are initialized to their specified values in design JSON.

> Second, indeed I create the children in constructor (the slider knob in the sliders asset) and it seems to work fine. I was mimicking how it's done in TCastleScrollView: https://github.com/castle-engine/castle-engine/blob/master/src/ui/castlecontrols_scrollview.inc#L598 - Though I already see that I didn't set FScrollArea.SetSubComponent(true);: looking through CGE documentation doesn't give much information:

Ah, SetSubComponent is kind of a 3rd approach, compared to 2 approaches I described so far in https://castle-engine.io/custom_components#_children_components . It allows to make the component non-internal (user can edit the properties) but at the same time user cannot select or delete the subcomponent. So it's kind of a middle ground between 2 approaches for children I described on https://castle-engine.io/custom_components#_children_components so far.

So... I just extended https://castle-engine.io/custom_components#_children_components . It now lists 3 useful cases how to setup children:

* 9.1. Internal children components (using SetTransient)
* 9.2. Children components that user can freely edit and remove (using TRegisteredComponent.OnCreate)
* 9.3. Children components on which user can edit properties (using SetSubComponent)

> I reimplemented InternalProposeName for my goals (if I remember correctly, because it was crashing), but then rolled the change back when I saw that having the "static" name in ScrollArea is ok and it seems like not causing any troubles in my usecase (but of course doesn't look clean)

I'd need to see a testcase to tell why it was crashing :) `InternalProposeName` is used a lot during CGE components editing. And I would very discourage from reimplementing it, it contains quite a few hardcoded cases that we may want to extend in the future.

**Comment by michaliskambi** [comment]
@eugeneloza Getting back to this issue -- can you reproduce it still, with latest CGE? 

I retested now, and (just like in the past, https://github.com/castle-engine/castle-engine/issues/411#issuecomment-1317635109 ) I cannot reproduce it.

**Comment by michaliskambi** [comment]
@eugeneloza I see this old issue is still waiting -- but I was never able to reproduce it. Is this still an issue for you? 

If you can reproduce, a new idea: test whether you can reproduce it with Qt backend (`castle-engine-qt5`  binary from precompiled CGE downloads) too.

**Comment by eugeneloza2factor** [comment]
Yes, unfortunately I can still reproduce the issue with the latest git version of CGE.

https://github.com/castle-engine/castle-engine/assets/73470521/cd139441-f48d-4ea1-a5cd-eb2f18038b52

I'll try QT version.

**Comment by eugeneloza2factor** [comment]
Unfortunately QT version didn't work for me :) Said that it has different plans for my weekends and I definitely have to upgrade Debian installation. Maybe when I'll finally get 2 days to reinstall everything the original issue will also get magically fixed?
![2024-02-24-185452_1366x768_scrot](https://github.com/castle-engine/castle-engine/assets/73470521/6609025f-13a8-4776-9cb9-00260215141a)
